components/brave_rewards/resources/tip/lib/host.ts

```diff
@@ -0,0 +1,225 @@
+  for (const [name, listener] of Object.entries(listeners)) {
+    self.cr.addWebUIListener(name, listener)
+  }
+}
+
+export function createHost (): Host {
+  const stateManager = createStateManager<HostState>({})
+  const dialogArgs = getDialogArgs()
+
+  addWebUIListeners({
+
+    rewardsInitialized () {
+      const { publisherKey } = dialogArgs
+
+      if (!publisherKey) {
+        stateManager.update({
+          hostError: { type: 'INVALID_DIALOG_ARGS' }
+        })
+        return
+      }
+
+      chrome.send('getRewardsParameters')
+      chrome.send('fetchBalance')
+      chrome.send('getReconcileStamp')
+      chrome.send('getExternalWallet', ['uphold'])
+      chrome.send('getOnlyAnonWallet')
+      chrome.send('getRecurringTips')
+      chrome.send('getPublisherBanner', [publisherKey])
+    },
+
+    externalWalletUpdated (externalWalletInfo: ExternalWalletInfo) {
+      stateManager.update({ externalWalletInfo })
+    },
+
+    publisherBannerUpdated (publisherInfo: PublisherInfo) {
+      stateManager.update({ publisherInfo })
+    },
+
+    rewardsParametersUpdated (rewardsParameters: RewardsParameters) {
+      stateManager.update({ rewardsParameters })
+    },
+
+    recurringTipsUpdated (tips?: RecurringTipInfo[]) {
+      if (!tips) {
+        return
+      }
+      for (const tip of tips) {
+        if (tip.publisherKey === dialogArgs.publisherKey) {
+          stateManager.update({
+            currentMonthlyTip: Number(tip.amount) || 0
+          })
+          break
+        }
+      }
+    },
+
+    reconcileStampUpdated (stamp: number) {
+      stateManager.update({ nextReconcileDate: new Date(stamp * 1000) })
+    },
+
+    balanceUpdated (result: { status: number, balance: BalanceInfo }) {
+      if (result.status === 0) {
+        stateManager.update({
+          balanceInfo: result.balance
+        })
+      } else {
+        stateManager.update({
+          hostError: {
+            type: 'ERROR_FETCHING_BALANCE',
+            code: result.status
+          }
+        })
+      }
+    },
+
+    onlyAnonWalletUpdated (onlyAnonWallet: boolean) {
+      stateManager.update({ onlyAnonWallet })
+    },
+
+    recurringTipRemoved (success: boolean) {
+      chrome.send('getRecurringTips')
+    },
+
+    recurringTipSaved (success: boolean) {
+      chrome.send('getRecurringTips')
+    },
+
+    reconcileCompleted (data: { type: number, result: number }) {
+      if (data.result === 0) {
+        chrome.send('fetchBalance')
+      }
+    }
+
+  })
+
+  self.i18nTemplate.process(document, self.loadTimeData)
+
+  chrome.send('dialogReady')
+
+  return {
+
+    get state () {
+      return stateManager.getState()
+    },
+
+    getString (key: StringKey) {
+      return self.loadTimeData.getString(key)
+    },
+
+    getDialogArgs () {
+      return dialogArgs
+    },
+
+    closeDialog () {
+      chrome.send('dialogClose')
+    },
+
+    processTip (amount: number, kind: TipKind) {
```

Comment:
Maybe `TipFrequency` is a more descriptive type name? Not sure... Maybe that would be too limiting.

Reply:
I'll leave the type name as `TipKind` for now - easy to change later if we need to.